src/python/pants/build_graph/build_configuration.py

```diff
@@ -136,29 +141,29 @@ def optionables(self):
         """
         return self._optionables
 
-    def register_rules(self, rules):
-        """Registers the given rules.
-
-        param rules: The rules to register.
-        :type rules: :class:`collections.Iterable` containing
-                     :class:`pants.engine.rules.Rule` instances.
-        """
+    def register_rules(
+        self, rules: Iterable[Union[Callable, Rule, UnionRule, PluginField]]
+    ) -> None:
         if not isinstance(rules, Iterable):
             raise TypeError("The rules must be an iterable, given {!r}".format(rules))
 
+        actual_rules: OrderedSet[Union[Callable[..., Any], Rule]] = OrderedSet()
+        for rule in rules:
```

Comment:
This works with `SubsystemRule` because it fits the `FunctionType`?

https://github.com/pantsbuild/pants/blob/31e41b1f661de7f0c6d745e9439fe7819e977cb1/src/python/pants/engine/rules.py#L182

Reply:
No. It's because SubsystemRule returns a TaskRule which is a Rule. It's the returned TaskRule that's in the rules list.